Muon double pulse analysis (#28830)
mantidproject/mantid: docs/source/algorithms/DoublePulseFit-v1.rst

.. algorithm::

.. summary::

.. relatedalgorithms::

.. properties::

Description
-----------

This algorithm is is designed to analyse double pulse Muon data. It fits two copies of the input function seperated by a time offset. The function fitted is of the form

.. math::

>   f(t) = N_1 f(t + t_s)_{input} + N_2 f(t)_{input}.

Reviewer: ```suggestion
  f(t) = N_1 f(t + t_s)_{input} + N_2 f(t)_{input},
```

> Where :math:`N_1` is the FirstPulseWeight, :math:`N_2` is the SecondPulseWeight and :math:`t_s` is the PulseOffset. The parameters of the two input functions are all tied. In practise this is achieved by taking a convolution of two delta functions with the input function.

Reviewer: ```suggestion
where :math:`N_1` is the FirstPulseWeight, :math:`N_2` is the SecondPulseWeight and :math:`t_s` is the PulseOffset. The corresponding parameters of the two input functions are tied to have the same value. In practise this is achieved by taking a convolution of two delta functions with the input function.
```

.. math::
  f(t) = f(t)_{input} * (N_1\delta(-ts) + N_2\delta(0)).

With the exception of the three new parameters PulseOffset, FirstPulseWeight and SecondPulseWeight the interface for DoublePulseFit is identical to that of Fit and the two can be used interchangeably.
The parameters and function output by this algorithm corrospond to the inner function only whilst the output fit curves corrospond to the full function.

> 

Reviewer: where is the code example/test case?

.. algorithm::

.. summary::

.. relatedalgorithms::

.. properties::

Description
-----------

This algorithm is is designed to analyse double pulse Muon data. It fits two copies of the input function seperated by a time offset. The function fitted is of the form

.. math::
  f(t) = N_1 f(t + t_s/2)_{input} + N_2 f(t - t_s/2)_{input},

where :math:`N_1` is the FirstPulseWeight, :math:`N_2` is the SecondPulseWeight and :math:`t_s` is the PulseOffset. The corresponding parameters of the two input functions are tied to have the same value. In practise this is achieved by taking a convolution of two delta functions with the input function;

.. math::
  f(t) = f(t)_{input} * (N_1\delta(-t_s/2) + N_2\delta(t_s/2)).

With the exception of the three new parameters PulseOffset, FirstPulseWeight and SecondPulseWeight the interface for DoublePulseFit is identical to that of Fit and the two can be used interchangeably.
The function output by this algorithm corrospond to the inner function only whilst the output fit curves corrospond to the full function.

Usage
-----

.. testcode::

  import numpy as np

>   from mantid.simpleapi import DoublePulseFit, CreateWorkspace, GausOsc, Fit

Reviewer: ```suggestion
  from mantid.simpleapi import DoublePulseFit, CreateWorkspace, GausOsc, Fit, AnalysisDataService
```

  # Create a workspace with two offset oscillations
  delta = 0.33
  x = np.linspace(0.,15.,100)
  x_offset = np.linspace(delta/2, 15. + delta/2, 100)
  x_offset_neg = np.linspace(-delta/2, 15. - delta/2, 100)

  testFunction = GausOsc(Frequency = 1.5, A=0.22)
  y1 = testFunction(x_offset_neg)
  y2 = testFunction(x_offset)
  y = y1/2+y2/2
  ws = CreateWorkspace(x,y)

  # Create functions to fit
  convolution =  FunctionFactory.createCompositeFunction('Convolution')
  innerFunction = FunctionFactory.createInitialized('name=GausOsc,A=0.2,Sigma=0.2,Frequency=1,Phi=0')
  deltaFunctions = FunctionFactory.createInitialized('(name=DeltaFunction,Height=0.5,Centre={},ties=(Height=0.5,Centre={});name=DeltaFunction,Height=0.5,Centre={},ties=(Height=0.5,Centre={}))'.format(-delta/2, -delta/2, delta/2, delta/2))
  convolution.setAttributeValue('FixResolution', False)
  convolution.add(innerFunction)
  convolution.add(deltaFunctions)
  innerFunctionSingle = FunctionFactory.createInitialized('name=GausOsc,A=0.2,Sigma=0.2,Frequency=1,Phi=0')
  
  DoublePulseFit(Function=innerFunctionSingle, InputWorkspace=ws,  PulseOffset = delta, StartX=0.0, EndX=15.0, Output='DoublePulseFit')

>   Fit(Function=convolution, InputWorkspace=ws, CreateOutput = True, StartX=0.0, EndX=15.0, Output='Fit', MaxIterations=1)

Reviewer: ```suggestion
  Fit(Function=convolution, InputWorkspace=ws, CreateOutput = True, StartX=0.0, EndX=15.0, Output='Fit')
```
to prove what its doing under the hood

  double_parameter_workspace = AnalysisDataService.retrieve('DoublePulseFit_Parameters')
  double_col_values = double_parameter_workspace.column(1)

  single_parameter_workspace = AnalysisDataService.retrieve('DoublePulseFit_Parameters')
  col_values = single_parameter_workspace.column(1)
  

>   print('Fitted value of A from DoublePulseFit is {}'.format(double_col_values[0]))

Reviewer: ```suggestion
  print('Fitted value of A from DoublePulseFit is {}'.format(round(double_col_values[0],2)))
```

>   print('Fitted value of Frequency from DoublePulseFit is {}'.format(double_col_values[2]))

Reviewer: ```suggestion
  print('Fitted value of Frequency from DoublePulseFit is {}'.format(round(double_col_values[2],2)))
```

>   print('Fitted value of A from Fit is {}'.format(col_values[0]))

Reviewer: ```suggestion
  print('Fitted value of A from Fit is {}'.format(round(col_values[0],2)))
```

>   print('Fitted value of Frequency from Fit is {}'.format(col_values[2]))

Reviewer: ```suggestion
  print('Fitted value of Frequency from Fit is {}'.format(round(col_values[2],2)))
```

.. math::
  f(t) = f(t)_{input} * (N_1\delta(-ts) + N_2\delta(0)).

With the exception of the three new parameters PulseOffset, FirstPulseWeight and SecondPulseWeight the interface for DoublePulseFit is identical to that of Fit and the two can be used interchangeably.
The function output by this algorithm corrospond to the inner function only whilst the output fit curves corrospond to the full function.

Usage
-----

.. code-block:: python

  import numpy as np
  from mantid.simpleapi import DoublePulseFit, CreateWorkspace, GausOsc
  from mantid.api import FunctionFactory

  # Create a workspace with two offset oscillations
  delta = 0.33
  x = np.linspace(0.,15.,100)
  x_offset = np.linspace(delta, 15. + delta, 100)
  testFunction = GausOsc(Frequency = 1.5, A=0.22)
  y1 = testFunction(x)
  y2 = testFunction(x_offset)
  y = y1+y2
  ws = CreateWorkspace(x,y)

  innerFunction = FunctionFactory.createInitialized('name=GausOsc,A=0.2,Sigma=0.2,Frequency=1,Phi=0')
  DoublePulseFit(Function=innerFunction, InputWorkspace=ws,  PulseOffset = delta, StartX=0.0, EndX=15.0, Output='DoublePulseFit')

> 

Reviewer: can we also check t he output parameters from the fit are roughly correct (f=1.5 and A=0.22)?

  y = y1+y2
  ws = CreateWorkspace(x,y)

  # Create functions to fit
  convolution =  FunctionFactory.createCompositeFunction('Convolution')
  innerFunction = FunctionFactory.createInitialized('name=GausOsc,A=0.2,Sigma=0.2,Frequency=1,Phi=0')

>   deltaFunctions = FunctionFactory.createInitialized('(name=DeltaFunction,Height=1,Centre=-0.33,ties=(Height=1,Centre=-0.33);name=DeltaFunction,Height=1,Centre=0,ties=(Height=1,Centre=0))')

Reviewer: should this not be `+/- 0.33/2` for each delta?

  convolution.setAttributeValue('FixResolution', False)
  convolution.add(innerFunction)
  convolution.add(deltaFunctions)
  innerFunctionSingle = FunctionFactory.createInitialized('name=GausOsc,A=0.2,Sigma=0.2,Frequency=1,Phi=0')
  
  DoublePulseFit(Function=innerFunctionSingle, InputWorkspace=ws,  PulseOffset = delta, StartX=0.0, EndX=15.0, Output='DoublePulseFit')
  Fit(Function=convolution, InputWorkspace=ws, CreateOutput = True, StartX=0.0, EndX=15.0, Output='Fit', MaxIterations=1)

  double_parameter_workspace = AnalysisDataService.retrieve('DoublePulseFit_Parameters')
  double_col_values = double_parameter_workspace.column(1)

>   single_parameter_workspace = AnalysisDataService.retrieve('DoublePulseFit_Parameters')

Reviewer: we are comparing the same table with itself. I have just tried running this and I get similar frequency values but not A

Reviewer: ah, just saw the late one.

Reviewer: ```suggestion
  single_parameter_workspace = AnalysisDataService.retrieve('Fit_Parameters')
```
